module/documents/actor/actor.mjs

```diff
@@ -1973,52 +2077,102 @@ export default class Actor5e extends Actor {
-    }
+    if ( !this.isOwner ) return ui.notifications.warn(game.i18n.localize("DND5E.PolymorphRevertWarn"));
 
-    // If we are reverting an unlinked token, simply replace it with the base actor prototype
+    /**
+     * A hook event that fires just before the actor is reverted to original form.
+     * @function dnd5e.transformActor
+     * @memberof hookEvents
+     * @param {Actor} this                 The original actor before transformation.
+     * @param {object} [options]
+     */
+    Hooks.callAll("dnd5e.revertOriginalForm", this, {renderSheet});
+    const previousActorIds = this.getFlag("dnd5e", "previousActorIds");
+    const isOriginalActor = !previousActorIds || previousActorIds.length <= 0;
+    const isRendered = this.sheet.rendered;
+
+    // Obtain a reference to the original actor
+    const original = game.actors.get(this.getFlag("dnd5e", "originalActor"));
+
+    // If we are reverting an unlinked token, grab the previous actorData, and create a new token
     if ( this.isToken ) {
-      const baseActor = game.actors.get(this.token.actorId);
-      const prototypeTokenData = await baseActor.getTokenData();
-      const tokenUpdate = {actorData: {}};
-      for ( let k of ["width", "height", "scale", "img", "mirrorX", "mirrorY", "tint", "alpha", "lockRotation", "name"] ) {
+      const baseActor = original ? original : game.actors.get(this.token.actorId);
+      if ( !baseActor ) {
+        ui.notifications.warn(game.i18n.format("DND5E.PolymorphRevertNoOriginalActorWarn", {
+          reference: this.getFlag("dnd5e", "originalActor")
+        }));
+        return;
+      }
+      const prototypeTokenData = await baseActor.getTokenDocument();
+      const actorData = foundry.utils.getProperty(this, "flags.dnd5e.previousActorData");
+      const tokenUpdate = this.token.toObject();
+      tokenUpdate.actorData = actorData ? actorData : {};
+
+      for ( const k of ["width", "height", "alpha", "lockRotation", "name"] ) {
         tokenUpdate[k] = prototypeTokenData[k];
       }
-      await this.token.update(tokenUpdate, {recursive: false});
+      for ( const k of ["offsetX", "offsetY", "scaleX", "scaleY", "src", "tint"] ) {
+        tokenUpdate.texture[k] = prototypeTokenData.texture[k];
+      }
+      tokenUpdate.sight = prototypeTokenData.sight;
+      tokenUpdate.detectionModes = prototypeTokenData.detectionModes;
+
       await this.sheet.close();
-      const actor = this.token.getActor();
-      actor.sheet.render(true);
-      return actor;
+      await canvas.scene?.deleteEmbeddedDocuments("Token", [this.token._id]);
+      const token = await TokenDocument.implementation.create(tokenUpdate, {
+        parent: canvas.scene, keepId: true, render: true
+      });
+      if ( isOriginalActor ) {
+        await this.unsetFlag("dnd5e", "isPolymorphed");
+        await this.unsetFlag("dnd5e", "previousActorIds");
+      }
+      if ( isRendered && renderSheet ) token.actor.sheet?.render(true);
```

Comment:
Rare, but this came up in testing where I would transform an actor, then drag its transformed version onto the canvas, then revert the original actor, and then attempt to revert the token. It worked well but threw an error here because (obviously) I had deleted the original actor.

```suggestion
      if ( isRendered && renderSheet ) token.actor?.sheet?.render(true);
```